### PR TITLE
Optimize remotecall pattern used in distributed loading

### PR DESCRIPTION
**Pull request by amitmurthy**
Came across this possible optimization while I was looking at the messages sent between workers in https://github.com/JuliaLang/julia/pull/21718

Currently the fetch/wait calls are 1) executed serially and 2) are an extra network call as compared with using a `@sync for p in pids; @async remotecall_fetch....` pattern.

Have replaced the `@spawnat followed by fetch/wait` calls with an async `remotecall_fetch`

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -173,18 +173,23 @@ function _require_from_serialized(node::Int, mod::Symbol, path_to_try::String, t
...
-                        catch ex
-                            isa(ex, Exception) ? ex : ErrorException(string(ex))
-                        end
-                    isa(m, Exception) ? m : nothing
-                end))
-            for p in others ]
-        for (id, ref) in refs
-            m = fetch(ref)
+
+        results = Any[]
```
Comment (alyst) [inline base/loading.jl]: `sizehint!(Vector{Tuple{Int,Any}}(), nprocs())`?

Looks like `others` is not required, should be ok to just skip `myid()` before `@async` block. One allocation less :)

**Comment by vtjnash** [comment]
I'm not sure I follow. Why is `@sync for p in pids; @async remotecall_fetch....` more efficient than `fetch.([@async remotecall.... for p in pids])`? Is the idea that the former runs the task management locally, whereas the latter runs it distributed?

**Comment by amitmurthy** [comment]
for pids (2,3,4,5) : 
`@sync for p in pids; @async remotecall_fetch....` results in
```
Message remotecall_fetch from 1 to 2
Message remotecall_fetch from 1 to 3
Message remotecall_fetch from 1 to 4
Message remotecall_fetch from 1 to 5

# responses can come back in any order

Message remotecall_fetch_response from 3 to 1
Message remotecall_fetch_response from 2 to 1
Message remotecall_fetch_response from 4 to 1
Message remotecall_fetch_response from 5 to 1
```

`fetch.([@async remotecall.... for p in pids])` results in 
```
Message remotecall from 1 to 2
Message remotecall from 1 to 3
Message remotecall from 1 to 4
Message remotecall from 1 to 5

# The above calls return a future that are then "fetched" serially

Message fetch from 1 to 2
Message fetch_response from 2 to 1
Message fetch from 1 to 3
Message fetch_response from 3 to 1
Message fetch from 1 to 4
Message fetch_response from 4 to 1
Message fetch from 1 to 5
Message fetch_response from 5 to 1
```

In the first case, the results of the `remotecall_fetch` are pushed asynchronously from the worker nodes. In the second case they are fetched serially.

**Comment by amitmurthy** [comment]
Will merge after a green CI

**Comment by tkelman** [comment]
why the rush? it's been open less than a day, give people a chance to see it

**Comment by amitmurthy** [comment]
Sure, no issues.

File: base/loading.jl [hunk 2]
```diff
@@ -460,8 +466,13 @@ function _require(mod::Symbol)
                 eval(Main, :(Base.include_from_node1($path)))
 
                 # broadcast top-level import/using from node 1 (only)
-                refs = Any[ @spawnat p eval(Main, :(Base.include_from_node1($path))) for p in filter(x -> x != 1, procs()) ]
-                for r in refs; wait(r); end
+                @sync begin
+                    for p in filter(x -> x != 1, procs())
+                        @async remotecall_fetch(p) do
```
Comment (JeffBezanson) [inline base/loading.jl]: Should use `remotecall_wait`, since the old version called `wait`.

Comment (amitmurthy) [inline base/loading.jl]: It is better to use a `remotecall_fetch` (after ensuring that the remote execution returns `nothing`) when we we do not need the result of the remote computation. Will avoid unnecessary storage and and clean-up of the result on the remote node.

**Comment by amitmurthy** [comment]
Will merge in a short while.

**Pull request merged**